maro/simulator/scenarios/vm_scheduling/virtual_machine.py

```diff
@@ -0,0 +1,61 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from typing import List
+
+
+class VirtualMachine:
+    """VM object.
+
+    The VM lifetime equals to the deletion tick - creation tick.
+    For example:
+        A VM's cpu utilization is {tick: cpu_utilization} = {0: 0.1, 1: 0.4, 2: 0.2}.
+        In our scenario, we define vm's creation tick as 0, and deletion tick as 3.
+        Its lifetime will be deletion tick - creation tick = 3 - 0 = 3
+
+    Args:
+        id (int): The VM id.
+        cpu_cores_requirement (int): The amount of virtual cores requested by VM.
+        memory_requirement (int): The memory requested by VM. The unit is (GBs).
+        lifetime (int): The lifetime of the VM, that is, deletion tick - creation tick + 1.
+    """
+    def __init__(self, id: int, cpu_cores_requirement: int, memory_requirement: int, lifetime: int):
+        # VM Requirement parameters.
+        self.id: int = id
+        self.cpu_cores_requirement: int = cpu_cores_requirement
+        self.memory_requirement: int = memory_requirement
+        self.lifetime: int = lifetime
+        # VM utilization list with VM cpu utilization(%) in corresponding tick.
+        self._utilization_series: List[float] = []
+        # The physical machine Id that the VM is assigned.
+        self.pm_id: int = -1
+        self._cpu_utilization: float = 0.0
+        self.creation_tick: int = -1
+        self.deletion_tick: int = -1
+
+    @property
+    def cpu_utilization(self) -> float:
+        return self._cpu_utilization
+
+    @cpu_utilization.setter
+    def cpu_utilization(self, cpu_utilization: float):
+        self._cpu_utilization = min(max(0, cpu_utilization), 100)
+
+    def get_utilization(self, cur_tick: int):
+        if cur_tick - self.creation_tick > len(self._utilization_series):
+            raise Exception(f"The tick {cur_tick} is invalid for the VM {self.id}.")
+
+        return self._utilization_series[cur_tick - self.creation_tick]
+
+    def add_utilization(self, cpu_utilization: float):
+        """VM CPU utilization list."""
+        # If cpu_utilization is smaller than 0, it means the missing data in the cpu readings file.
+        # TODO: We use the last utilization, it could be further refined to use average or others.
+        if cpu_utilization < 0.0:
+            self._utilization_series.append(self._utilization_series[-1])
+        else:
+            self._utilization_series.append(cpu_utilization)
+
+    def get_historical_utilization_series(self, cur_tick: int) -> List[float]:
```

Comment:
I think it's a fake protected interface, VM doesn't own cur_tick
Why we need this?

Reply:
Currently, outside the BE, agent can't directly access the VM instance.
Also, the VM instance doesn't host the "cur_tick" info. _But we are not sure do we need to add this?_

Since the VM historical utilization series not in snapshots, we design to add an interface for BE. So that agent can access this info through BE, like

```
def take_action(pending_decision_event):
    ...
    for VM in pending_decision_event.valid_pm[i]:
        vm_historical_utilization_series = BE.get_vm_series(vm_id)

def BE.get_vm_series(vm_id):  # not implemented yet
    return self._live_vm[vm_id].get_historical_utilization_series(self.cur_tick)
```

How about this?